unit_scaling/_modules.py

```diff
@@ -241,54 +443,111 @@ class TransformerLayer(nn.Module):
-        self.mhsa_layer_norm = LayerNorm(hidden_size)
-        self.mhsa = MHSA(hidden_size, heads, dropout_p, constraint=constraint)
-        self.mlp_layer_norm = LayerNorm(hidden_size)
+        self.mhsa_norm = RMSNorm(hidden_size)
+        self.mhsa = MHSA(
+            hidden_size,
+            heads,
+            is_causal=is_causal,
+            dropout_p=dropout_p,
+            constraint=constraint,
+        )
+        self.mlp_norm = RMSNorm(hidden_size)
         self.mlp = MLP(hidden_size, act_fn, constraint=constraint)
 
     def forward(self, input: Tensor) -> Tensor:
         input, skip = U.residual_split(input, tau=self.mhsa_tau)
-        input = self.mhsa_layer_norm(input)
+        input = self.mhsa_norm(input)
         input = self.mhsa(input)
         input = U.dropout(input, self.dropout_p, self.training)
         input = U.residual_add(input, skip, tau=self.mhsa_tau)
 
         input, skip = U.residual_split(input, tau=self.mlp_tau)
-        input = self.mlp_layer_norm(input)
+        input = self.mlp_norm(input)
         input = self.mlp(input)
         input = U.dropout(input, self.dropout_p, self.training)
         return U.residual_add(input, skip, tau=self.mlp_tau)
 
 
+@inherit_docstring(
+    "A :class:`torch.nn.Sequential` that automatically configures the depth"
+    " for sake of scaling."
+)
+class Trunk(nn.Sequential):
+    def __init__(self, *args: Any) -> None:
+        super().__init__(*args)
+        for name, parameter in self.named_parameters():
+            if not has_parameter_data(parameter):
+                raise ValueError(
+                    f"Parameter {name} is not a unit_scaling.Parameter."
+                    " Is it from a regular nn.Module?"
+                )
+            parameter.mup_scaling_depth = len(self)
+
+
+class TransformerStack(Trunk):
+    """A **unit-scaled** transformer stack, applying a residual scaling rule.
+
+    See :code:`TransformerLayer` for arguments.
+
+    Args:
+        layers (int): number of transformer layers.
+        residual_scaling (Callable[[int, int], float], optional): scheme for
+            controlling residual weights in the transformer trunk; see
+            :func:`unit_scaling.core.functional.transformer_residual_scaling_rule`
+            (default).
+    """
+
+    def __init__(
+        self,
+        layers: int,
+        residual_scaling: ResidualScalingFn = transformer_residual_scaling_rule(),
+        **kwargs: Any,
+    ) -> None:
+        super().__init__(
+            *(
+                TransformerLayer(
+                    **kwargs,
+                    mhsa_tau=residual_scaling(2 * i, 2 * layers),
+                    mlp_tau=residual_scaling(2 * i + 1, 2 * layers),
+                )
+                for i in range(layers)
+            )
+        )
+
+
 @format_docstring(variadic_constraint_docstring)
-class TransformerDecoder(nn.Module):  # pragma: no cover
+class TransformerDecoder(nn.Sequential):  # pragma: no cover
```

Comment:
Same problems using `nn.Sequential` here. We'd typically register tensor data as buffers in the root module that we want to reuse across layers, which is difficult with `nn.Sequential`

Reply:
Agree; we can switch to regular if/when we do this. TBH I'm not sure about such high-level modules "just for demonstrating scaling" continuing to live inside `uu.*`, but thought it worth keeping them for now, at least for sake of illustration.